### PR TITLE
Feature detail video

### PR DESCRIPTION
**Pull request by Giang-st1-00**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "react-hook-form": "^7.43.9",
     "react-icons": "^4.8.0",
     "react-modal": "^3.16.1",
+    "react-router-dom": "^6.13.0",
```
Comment (Duy-Tinh-Dev) [inline package.json]: delete

File: model/video.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+export interface IVideoPayload {
+  author: {
```
Comment (Duy-Tinh-Dev) [inline model/video.ts]: should separate author to a new Interface

File: model/video.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  sharers: number;
+  hashtags: Array<string>;
+  comments: Array<string>;
+  likers: number | string;
+  video_id_children: string;
+  parent_id: string;
+  duration: number;
+  break_point: number;
+  pathVideo: string;
+  fullVideoInfo?: any;
```
Comment (Duy-Tinh-Dev) [inline model/video.ts]: dont should using"any"

File: apis/video.ts [hunk 4]
```diff
@@ -2,7 +2,10 @@ import axiosClient from "./axiosClient";
 
 const videoApi = {
   getAllVideo: () => {
-    return axiosClient.get("/api/video");
+    return axiosClient.get("/api/video/");
+  },
+  getVideoById: (id: any) => {
```
Comment (Duy-Tinh-Dev) [inline apis/video.ts]: !any

File: components/common/StatusVideo/ControlsVideo/index.tsx [hunk 5]
```diff
@@ -0,0 +1,144 @@
+import { Dispatch, SetStateAction } from "react";
+import { FaPause, FaPlay } from "react-icons/fa";
+import { MdFullscreen } from "react-icons/md";
+
+interface IControlsVideo {
+  dataVideo: any;
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ControlsVideo/index.tsx]: !any

File: components/common/StatusVideo/ControlsVideo/index.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+interface IControlsVideo {
+  dataVideo: any;
+  totalTime: number;
+  statusVideo: string;
+  currentTime: number;
+  setCurrentTime: Dispatch<SetStateAction<number>>;
+  handlePlayByPlayer: () => void;
+  handlePauseByPlayer: () => void;
+  setTimePlayerModal: (value: any) => void;
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ControlsVideo/index.tsx]: !any

File: styles/globals.css [hunk 7]
```diff
@@ -35,4 +30,32 @@ input[type="text"]:not(:placeholder-shown) ~ .input-text {
...
+
+.videoModal_container:hover .reactVideo {
+  opacity: 100%;
+}
+
+.videoModal_container:hover .closeButton {
+  opacity: 100%;
+}
+
+/* .video-list:hover .control_bar {
```
Comment (Duy-Tinh-Dev) [inline styles/globals.css]: delete code not using

File: pages/index.tsx [hunk 8]
```diff
@@ -11,10 +11,17 @@ Index.Layout = MainLayout;
 export default Index;
 
 export async function getStaticProps() {
-  let posts;
+  let posts: any;
```
Comment (Duy-Tinh-Dev) [inline pages/index.tsx:14]: !any

File: pages/index.tsx [hunk 9]
```diff
@@ -11,10 +11,17 @@ Index.Layout = MainLayout;
...
 
   try {
     posts = await videoApi.getAllVideo();
+
+    for (let i = 0; i < posts.metaData.length; i++) {
+      const parentId = posts.metaData[i].parent_id;
+      const parentResponse: any = await videoApi.getVideoById(parentId);
```
Comment (Duy-Tinh-Dev) [inline pages/index.tsx:21]: !any

File: pages/video/[postId].tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
+import videoApi from "@/apis/video";
+import MainLayout from "@/components/layout/MainLayout";
+import DetailsVideo from "@/Views/DetailsVideo";
+
+export default function Post({ post }: any) {
```
Comment (Duy-Tinh-Dev) [inline pages/video/[postId].tsx:5]: fix all any this is page

File: Views/Home/index.tsx [hunk 11]
```diff
@@ -1,143 +1,15 @@
...
-  FaSwatchbook,
-  FaCommentAlt,
-  FaShare,
-} from "react-icons/fa";
-import { TiPlus } from "react-icons/ti";
-import { IoIosArrowDown } from "react-icons/io";
-import { AiTwotoneHeart } from "react-icons/ai";
-import { HiSparkles } from "react-icons/hi";
-import ModalVideo from "@/components/common/ModalVideo";
 import VideoStatusUsingAPI from "@/components/common/StatusVideo";
```
Comment (Duy-Tinh-Dev) [inline Views/Home/index.tsx]: +Should reman the same name folder
+i think should using a shorter name such as : "VideoShort"

File: components/common/StatusVideo/ReactVideo/index.tsx [hunk 12]
```diff
@@ -0,0 +1,90 @@
+import Image from "next/image";
+import { useState } from "react";
+import { AiTwotoneHeart } from "react-icons/ai";
+import { FaCommentAlt, FaShare } from "react-icons/fa";
+import { HiSparkles } from "react-icons/hi";
+import { TiPlus } from "react-icons/ti";
+
+interface IReactVideo {
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ReactVideo/index.tsx]: should has suffixes "props" such as "IReactVideoProps"

File: components/common/StatusVideo/ReactVideo/index.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+import { TiPlus } from "react-icons/ti";
+
+interface IReactVideo {
+  pathAvatar: string;
+  heartCount: string | number;
+  commentCount: string | number;
+  shareCount: string | number;
+}
+
+export default function ReactVideo({
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ReactVideo/index.tsx]: i think you should rename to "ActionVideo"

File: components/common/StatusVideo/index.tsx [hunk 14]
```diff
@@ -1,16 +1,30 @@
...
-export default function VideoStatusUsingAPI({ data }: any) {
-  const [status, setStatus] = useState("Ready");
+interface IVideoStatusUsingAPI {
+  data: IVideoShortPayload;
+}
+
+export default function VideoStatusUsingAPI({ data }: IVideoStatusUsingAPI) {
+  const [status, setStatus] = useState("NA");
+  const [statusModal, setStatusModal] = useState("NA");
   const [player, setPlayer] = useState<any>(null);
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/index.tsx]: !any

File: components/common/StatusVideo/index.tsx [hunk 15]
```diff
@@ -1,16 +1,30 @@
...
+  const [playerModal, setPlayerModal] = useState<any>(null);
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/index.tsx]: !any

File: components/common/StatusVideo/index.tsx [hunk 16]
```diff
@@ -31,32 +45,183 @@ export default function VideoStatusUsingAPI({ data }: any) {
       });
       setVideoTag(video);
     }
-  }, []);
+
+    const container_modal: any = modalVideoRef.current;
+    if (container_modal) {
+      const iframe_modal = container_modal.querySelector("iframe");
+      const isIframeExist_modal = iframe_modal !== null;
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/index.tsx]: move logic it to logic "if"
it also in line 39

File: components/common/StatusVideo/index.tsx [hunk 17]
```diff
@@ -31,32 +45,183 @@ export default function VideoStatusUsingAPI({ data }: any) {
...
+  const handleCloseModal = () => {
+    setOpenModalVideo(false);
+    if (window.history.length > 2) {
+      window.history.back();
+    }
+    setTimePlayer(currentTime - data.break_point);
+  };
+
+  //------------
+  useEffect(() => {
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/index.tsx]: move useEffect to the top

File: components/common/StatusVideo/index.tsx [hunk 18]
```diff
@@ -31,32 +45,183 @@ export default function VideoStatusUsingAPI({ data }: any) {
...
+      handlePauseByPlayer();
+    } else if (!isOpenModalVideo) {
+      handlePauseByPlayerModal();
+      handlePlayByPlayer();
+    }
+  }, [isOpenModalVideo]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  return (
+    <div ref={videoRef} className="videoModal_container w-full h-full">
+      {status != "NA" && (
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/index.tsx]: should using !==

File: components/common/StatusVideo/InformVideo/index.tsx [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+import { FaExternalLinkAlt, FaSwatchbook } from "react-icons/fa";
+import { IoIosArrowForward, IoIosArrowBack } from "react-icons/io";
+
+interface IInformVideo {
+  title: string;
+  caption: string;
+  hashtags: Array<string>;
+}
+
+export default function InformVideo({
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/InformVideo/index.tsx]: should using "DescriptionVideo"

File: components/common/StatusVideo/ModalVideo/index.tsx [hunk 20]
```diff
@@ -0,0 +1,92 @@
+import { IVideoPayload } from "@/model/video";
+import { Dispatch, MutableRefObject, SetStateAction, useEffect } from "react";
+import { CgClose } from "react-icons/cg";
+import ControlsVideo from "../ControlsVideo";
+import InformVideo from "../InformVideo";
+import ReactVideo from "../ReactVideo";
+
+interface IModalVideo {
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ModalVideo/index.tsx]: IModalVideoProps

File: components/common/StatusVideo/ControlsVideo/index.tsx [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+
+  return (
+    <div className="control_bar translate-y-[70%] transition duration-500 ease-in-out w-full absolute bottom-0">
+      <div className="py-1 px-4">
+        <div
+          className="hover:cursor-pointer"
+          onClick={(e) => handleProgressBarClick(e)}
+        >
+          <div className="relative bg-opacity-50 flex items-center justify-between">
+            {Array.isArray(dataVideo) ? (
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ControlsVideo/index.tsx]: should using "id" of api response to set field "key" in function map

File: components/common/StatusVideo/ControlsVideo/index.tsx [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+              className="absolute bg-[#fff] h-[10px] w-[10px] rounded-full hover:cursor-pointer translate-x-[-50%]"
+            ></div>
+          </div>
+        </div>
+      </div>
+      <div className="px-5 h-[50px] box-border flex justify-between items-center text-white">
+        <div
+          className="p-2 hover:cursor-pointer"
+          onClick={() => {
+            if (statusVideo == "Playing") {
```
Comment (Duy-Tinh-Dev) [inline components/common/StatusVideo/ControlsVideo/index.tsx]: should create a new function above
using operator === instead ==

**Comment by lamxuanthinh** [comment]
Screenshot 2023-07-01 at 13 04 02
Please resolve this conflict @Giang-st1-00

**Comment by kennytruong06** [comment]
oki

**Comment by lamxuanthinh** [comment]
@Giang-st1-00  Are you sure you want me to merge this branch into the `develop` branch

**Comment by Giang-st1-00** [comment]
yes

**Pull request merged**